proposals/params-span.md

```diff
@@ -0,0 +1,138 @@
+Otherwise the array will be allocated on the heap.
+
+```csharp
+Span<int> s = new[] { i, j, k };   // int[] on the stack
+WriteLine(fmt, new[] { x, y, z }); // object[] on the stack for WriteLine(string fmt, ReadOnlySpan<object> args);
+```
+
+### Array re-use
+The compiler _may_ reuse an implicitly allocated array across multiple uses within a single thread executing a method:
+- At the same call-site (within a loop) or
+- At distinct call-sites if the lifetime of the spans do not overlap, and the array length is sufficient, and
+  - the element types are managed types that are considered identical by the runtime, or
+  - the element types are unmanaged types of the same size.
+
+An implicitly allocated array may be reused regardless of whether the array was created on the stack or the heap.
+
+### Lowering implicit allocation
+For an array `T[]` of constant length `N`, the compiler will synthesize a `struct` with `N` fields of type `T` where the layout and alignment of the fields matches the alignment of elements in `T[]`.
+The array will be represented with a value of the synthesized `struct`, and a `Span<T>` created from a `ref` to the first field.
+
+For example, a call to `Console.WriteLine(fmt, x, y, z);` would be emitted as:
+```csharp
+[StructLayout(LayoutKind.Sequential)]
+internal struct __ValueArray3<T> { public T Item1, Item2, Item3; };
+
+var values = new __ValueArray3<object>() { Item1 = x, Item2 = y, Item3 = z };
+var span = MemoryMarshal.CreateSpan(ref values.Item1, 3);
+Console.WriteLine(fmt, (ReadOnlySpan<object>)span); // WriteLine(string format, params ReadOnlySpan<object?> arg)
+```
+
+The synthesized `struct` type will be `internal`, with an unspeakable name.
+
+Since `T` is a valid type argument (and used in `Span<T>`), the synthesized type will be a constructed generic type where the underlying generic type is shared across other call sites in the compilation with the same number of arguments.
+
+## Open issues
+### Is `params Span<T>` necessary?
+Is there a reason to support `params` parameters of type `Span<T>` in addition to `ReadOnlySpan<T>`? Is allowing mutation within the `params` method useful?
+
+### Is `params IEnumerable<T>` necessary?
+If the compiler allows `params ReadOnlySpan<T>`, then new APIs that require `params` could use `params ReadOnlySpan<T>` instead of `params T[]` because `T[]` is implicitly convertible to `ReadOnlySpan<T>`. And even legacy APIs could add a `params ReadOnlySpan<T>` overload where the existing `params T[]` simply delegates to the new overload.
+
+There is no conversion from `IEnumerable<T>` to `ReadOnlySpan<T>` however, so allowing `params IEnumerable<T>` is essentially asking APIs to provide two overloads for `params` methods: `params ReadOnlySpan<T>` and `params IEnumerable<T>`. 
+
+Are scenarios for `params IEnumerable<T>` sufficiently compelling to justify that?
```

Comment:
yes, i believe so.  This feature is not jsut about perf, but about ergonomics.  The requirement to take in an array has been very limiting to many apis, requiring an array overload (for the ease of passing values), and an IEnumerable/IList/etc. overload (to allow other collections to be passed in).   It's been a very longstanding request from teh community that this sort of common pattern not require multiple overloads when one would be clear and sufficient.

Reply:
Could the compiler generate a static instance of an ImmutableArray<T> and share it across all calls to `params IEnumerable<T>`?

Reply:
No because in the case of anything but `Span<T>` type we must assume the implementation smuggles away the argument. Hence re-use is not safe.

Reply:
But if the backing type is an ImmutableArray that shouldn't matter? I'm taking about a case where the params are constants.

Reply:
In that specific case yes we could put it in an `ImmutableArray<T>`. I'm not sure it's worth it. 

- Forces the compiler to depend on a new type and assign specific semantics to it. Generally we only do that for special types. 
- It's not a total win, there is still allocations at the call site (boxing the `ImmutableArray<T>`) and at the callee (allocation of `IEnumerator` during enumeration). 

Don't think we'll save enough for it to be worth the complexity.

Reply:
You called store the ImmutableArray already boxed as an IEnunerable to save on the first allocation. But agreed, less than perfect.